cpp/src/sampling/detail/sampling_utils_impl.cuh

```diff
@@ -448,5 +626,455 @@ sample_edges(
+              handle.get_thrust_policy(),
+              labels->begin(),
+              labels->end(),
+              thrust::make_zip_iterator(
+                majors.begin(), minors.begin(), edge_types->begin(), hops->begin()));
+          } else {
+            thrust::sort_by_key(
+              handle.get_thrust_policy(),
+              labels->begin(),
+              labels->end(),
+              thrust::make_zip_iterator(majors.begin(), minors.begin(), edge_types->begin()));
+          }
+        } else {
+          if (hops) {
+            thrust::sort_by_key(
+              handle.get_thrust_policy(),
+              labels->begin(),
+              labels->end(),
+              thrust::make_zip_iterator(majors.begin(), minors.begin(), hops->begin()));
+          } else {
+            thrust::sort_by_key(handle.get_thrust_policy(),
+                                labels->begin(),
+                                labels->end(),
+                                thrust::make_zip_iterator(majors.begin(), minors.begin()));
+          }
+        }
+      }
+    }
+
+    size_t num_unique_labels =
+      thrust::count_if(handle.get_thrust_policy(),
+                       thrust::make_counting_iterator<size_t>(0),
+                       thrust::make_counting_iterator<size_t>(labels->size()),
+                       [d_labels = labels->data()] __device__(size_t idx) {
+                         return (idx == 0) || (d_labels[idx - 1] != d_labels[idx]);
+                       });
+
+    rmm::device_uvector<label_t> unique_labels(num_unique_labels, handle.get_stream());
+
+    thrust::copy_if(
+      handle.get_thrust_policy(),
+      thrust::make_zip_iterator(thrust::make_counting_iterator<size_t>(0), labels->begin()),
+      thrust::make_zip_iterator(thrust::make_counting_iterator<size_t>(labels->size()),
+                                labels->end()),
+      thrust::make_zip_iterator(thrust::make_discard_iterator(), unique_labels.begin()),
+      [d_labels = labels->data()] __device__(auto tuple) {
+        size_t idx = thrust::get<0>(tuple);
+        return (idx == 0) || (d_labels[idx - 1] != d_labels[idx]);
+      });
+
+    thrust::transform(handle.get_thrust_policy(),
+                      labels->begin(),
+                      labels->end(),
+                      labels->begin(),
+                      [d_unique_labels = raft::device_span<label_t const>{
+                         unique_labels.data(), unique_labels.size()}] __device__(label_t label) {
+                        auto pos = thrust::lower_bound(
+                          thrust::seq, d_unique_labels.begin(), d_unique_labels.end(), label);
+                        return static_cast<label_t>(thrust::distance(d_unique_labels.begin(), pos));
+                      });
+
+    offsets = detail::compute_sparse_offsets<size_t>(
+      labels->begin(), labels->end(), size_t{0}, unique_labels.size(), handle.get_stream());
+    labels = std::move(unique_labels);
```

Comment:
Can't you just use `reduce_by_key`? `labels` are already sorted and you want to compute the counts of each labels (you can compute offsets from counts by running scan).

Reply:
Refactored in the next push to look like the linked renumbering code.